### PR TITLE
Fix trigger kwarg encryption migration

### LINKED ISSUE #38836: DB migrate throws na error on encrypt_trigger_kwargs
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.9.0

### What happened?

`airflow db migrate` throws an error:

```
[2024-04-08T13:41:38.179+0000] {migration.py:216} INFO - Context impl PostgresqlImpl.                                                                                   │
[2024-04-08T13:41:38.180+0000] {migration.py:219} INFO - Will assume transactional DDL.                                                                                 │
[2024-04-08T13:41:38.185+0000] {db.py:1650} INFO - Creating tables                                                                                                      │
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.                                                                                                          │
INFO  [alembic.runtime.migration] Will assume transactional DDL.                                                                                                        │
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.                                                                                                          │
INFO  [alembic.runtime.migration] Will assume transactional DDL.                                                                                                        │
Traceback (most recent call last):                                                                                                                                      │
  File "/usr/local/bin/airflow", line 8, in <module>                                                                                                                    │
    sys.exit(main())                                                                                                                                                    │
  File "/usr/local/lib/python3.10/site-packages/airflow/__main__.py", line 58, in main                                                                                  │
    args.func(args)                                                                                                                                                     │
  File "/usr/local/lib/python3.10/site-packages/airflow/cli/cli_config.py", line 49, in command                                                                         │
    return func(*args, **kwargs)                                                                                                                                        │
  File "/usr/local/lib/python3.10/site-packages/airflow/utils/cli.py", line 114, in wrapper                                                                             │
    return f(*args, **kwargs)                                                                                                                                           │
  File "/usr/local/lib/python3.10/site-packages/airflow/utils/providers_configuration_loader.py", line 55, in wrapped_function                                          │
    return func(*args, **kwargs)                                                                                                                                        │
  File "/usr/local/lib/python3.10/site-packages/airflow/cli/commands/db_command.py", line 130, in migratedb                                                             │
    db.upgradedb(                                                                                                                                                       │
  File "/usr/local/lib/python3.10/site-packages/airflow/utils/session.py", line 79, in wrapper                                                                          │
    return func(*args, session=session, **kwargs)                                                                                                                       │
  File "/usr/local/lib/python3.10/site-packages/airflow/utils/db.py", line 1674, in upgradedb                                                                           │
    encrypt_trigger_kwargs(session=session)                                                                                                                             │
  File "/usr/local/lib/python3.10/site-packages/airflow/utils/db.py", line 982, in encrypt_trigger_kwargs                                                               │
    trigger.kwargs = BaseSerialization.deserialize(json.loads(trigger.encrypted_kwargs))                                                                                │
  File "/usr/local/lib/python3.10/json/__init__.py", line 346, in loads                                                                                                 │
    return _default_decoder.decode(s)                                                                                                                                   │
  File "/usr/local/lib/python3.10/json/decoder.py", line 337, in decode                                                                                                 │
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())                                                                                                                   │
  File "/usr/local/lib/python3.10/json/decoder.py", line 355, in raw_decode                                                                                             │
    raise JSONDecodeError("Expecting value", s, err.value) from None                                                                                                    │
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

### What you think should happen instead?

no error

### How to reproduce

use python:3.10.12-slim-bookworm image with airflow 2.9.0. deploy into kubernetes cluster 1.26.5. Celery Executor. run `airflow db migrate`

### Operating System

Debian GNU/Linux 12 (bookworm)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==8.19.0
apache-airflow-providers-celery==3.6.1
apache-airflow-providers-cncf-kubernetes==8.0.1
apache-airflow-providers-common-io==1.3.0
apache-airflow-providers-common-sql==1.11.1
apache-airflow-providers-fab==1.0.2
apache-airflow-providers-ftp==3.7.0
apache-airflow-providers-http==4.10.0
apache-airflow-providers-imap==3.5.0
apache-airflow-providers-postgres==5.10.2
apache-airflow-providers-smtp==1.6.1
apache-airflow-providers-sqlite==3.7.1

### Deployment

Other Docker-based deployment

### Deployment details

on-premise Kubernetes cluster 1.26.5, celery executor, PostgreSQL 13.13

### Anything else?

Related: https://github.com/apache/airflow/commit/6f0e5bce590de567d5010220d2ca392b22630a7b #38358 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by jedcunningham**
Closes: #38836 
Alternative to #38876 

Do the encryption in the migration itself, and fix support for offline migrations as well.

The offline up migration won't actually encrypt the trigger kwargs as there isn't a safe way to accomplish that, so the decryption processes checks and short circuits if it isn't encrypted.

The offline down migration will now print out a warning that the offline migration will fail if there are any running triggers. I think this is the best we can do for that scenario (and folks willing to do offline migrations will hopefully be able to understand the situation).

This also solves the "encrypting the already encrypted kwargs" bug in 2.9.0.

### CODE REVIEW

File: airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py
Comment (jedcunningham) [inline airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:70]: I don't think we can do any better than spitting out a warning :(.

But if folks have ideas here, happy to hear them.

File: airflow/models/trigger.py
Comment (jedcunningham) [inline airflow/models/trigger.py:121]: We could start doing this if we'd like to, but I'd suggest we do it in a follow up PR.

File: airflow/models/trigger.py
Comment (jedcunningham) [inline airflow/models/trigger.py:123]: This fixes the offline upgrade path by detecting the kwargs aren't actually encrypted.

File: tests/models/test_trigger.py
Comment (jedcunningham) [inline tests/models/test_trigger.py]:
```suggestion
```

Nothing to see here 🤦

File: tests/models/test_trigger.py
Comment (jedcunningham) [inline tests/models/test_trigger.py]:
```suggestion
    # force the `encrypted_kwargs` to be unencrypted, like they would be after an offline upgrade
```

File: airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py
Comment (uranusjr) [reply airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:70]: Instead of emitting a warning, would it be possible to detect the case and just fail right here? This would make recovery a bit easier.

Comment (jedcunningham) [reply airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:70]: It's a moving target. There may be none now but when they go run the generated sql it blows up.

Comment (jedcunningham) [reply airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:70]: This is what it looks like in the generated sql:

```
BEGIN;

-- Running downgrade 1949afb29106 -> ee1467d4aa35


------------
--  WARNING: Unable to decrypt trigger kwargs automatically in offline mode!
--  If any trigger rows exist when you do an offline downgrade, the migration will fail.
------------

ALTER TABLE trigger ALTER COLUMN kwargs TYPE JSON USING kwargs::json;

UPDATE alembic_version SET version_num='ee1467d4aa35' WHERE alembic_version.version_num = '1949afb29106';


```

File: airflow/models/trigger.py
Comment (Lee-W) [inline airflow/models/trigger.py:123]: I think it's unlikely 🤔 but is it possible the first letter after encryption is `{`.

Comment (Lee-W) [reply airflow/models/trigger.py:123]: Maybe we can catch the exception and run the else block below in that case?

Comment (jedcunningham) [reply airflow/models/trigger.py:123]: The [result is base64](https://cryptography.io/en/latest/fernet/#cryptography.fernet.Fernet.encrypt), so no risk. But if you want to refactor to decrypt/catch instead, have at it 👍

Comment (Lee-W) [reply airflow/models/trigger.py:123]: If there's no risk, I think it's better keep it as it is for now to avoid unnecessary logic 👍

File: airflow/models/trigger.py
Comment (uranusjr) [reply airflow/models/trigger.py:121]: Let’s add an issue to track this.

**Review by uranusjr — APPROVED** [review]

**Review by Lee-W — APPROVED** [review]

**Comment by jedcunningham** [comment]
We've tested the matrix of postges/mysql/sqlite and online/offline migrations. Didn't find any issues 👍.

I think this should be good. I'd like to get 1 more approval on this before we merge it though.

File: airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py
Comment (dstandish) [inline airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:62]: i'm fuzzy on this... but are all upgrades supposed to be in a single transaction?

File: airflow/models/trigger.py
Comment (jedcunningham) [reply airflow/models/trigger.py:121]: Done: #39265

File: airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py
Comment (dstandish) [inline airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:55]: very much a style thing, but i think doing a pre-emptive return better reveals the intention in this case

i.e. `if context.is_offline_mode(): return` but feel free to disregard

sorta just makes it easier to see what "the main migration" is ... indented makes it look like it's conditional

File: airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py
Comment (jedcunningham) [reply airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:62]: Yeah, this is something I was a little suspect of also, but haven't had a chance to verify it yet. Let me do that now.

That said, this is a pattern we have in existing migrations already 🤷.

File: airflow/models/trigger.py
Comment (dstandish) [reply airflow/models/trigger.py:123]: FWIW i kindof like EAFP better.... i guess i'm using that acronym now...

File: airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py
Comment (jedcunningham) [reply airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:62]: So turns out this doesn't actually commit the DDL OR DML, the "real" commit happens when alembic does it in a higher scope. Should be good here.

Comment (dstandish) [reply airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:62]: weird. i guess it's a different session.

File: airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py
Comment (jedcunningham) [reply airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:55]: Yep. I generally like doing exactly that, not sure why I didn't this time around.

At this point, I'm going to save it for a future refactor PR though.

File: airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py
Comment (jedcunningham) [reply airflow/migrations/versions/0140_2_9_0_update_trigger_kwargs_type.py:62]: Yeah it behaves like a "child session", which is kinda cool.

**Review by dstandish — APPROVED** [review]

**Pull request merged**
